File: cpp/40_数组中只出现一次的数字.cpp
#include <iostream>
#include <vector>
#include <map>
#include <stdio.h>
#include <sys/time.h>

class Solution {
public:
    // 常规思路：使用map，时间复杂度O(2n)，空间复杂度O(n)
    void FindNumsAppearOnce1(std::vector<int> data, int* num1, int* num2) {
        if (data.empty()) {
            return;
        }

        std::map<int, int> map;
        for (auto i : data) {
            if (map.count(i) == 0) {
                map[i] = 1;
            } else {
                map[i]++;
            }
        }

        std::vector<int> tmp;
        for (auto i : data) {
            if (map[i] == 1) {
                tmp.push_back(i);
            }
        }

        *num1 = tmp[0];
        *num2 = tmp[1];
    }

    // 高效方式：使用异或，时间复杂度O(2n)，空间复杂度O(1)
    void FindNumsAppearOnce(std::vector<int> data, int* num1, int* num2) {
        if (data.empty()) {
            return;
        }

        int diff = 0;
        for (auto i : data) {
            diff ^= i;
        }
        diff &= -diff;

        *num1 = 0;
        *num2 = 0;
        for (auto i : data) {
            if ((i & diff) == 0) { // 注意运算符优先级
                *num1 ^= i;
            } else {
                *num2 ^= i;
            }
        }
    }

    // 测试函数
    void test(std::vector<int>& data) {
        int num1, num2;
        struct timeval start, end;
        printf("=====\n");
        for (auto func : this->func_vec_) {
            gettimeofday(&start, nullptr);
            (this->*func)(data, &num1, &num2);
            gettimeofday(&end, nullptr);
            printf("result: [%d, %d], time(us): %ld\n", num1, num2, end.tv_usec - start.tv_usec);
        }
    }

private:
    typedef void (Solution::*func_ptr)(std::vector<int>, int*, int*);
    std::vector<func_ptr> func_vec_ = {&Solution::FindNumsAppearOnce1,
                                       &Solution::FindNumsAppearOnce};
};

int main(int argc, char* argv[])
{
    std::vector<int> data = {2, 4, 3, 6, 3, 2, 5, 5};
    std::vector<int> data2 = {1, 2};

    Solution s;
    s.test(data);
    s.test(data2);

    return 0;
}

File: python/40_数组中只出现一次的数字.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

import time

class Solution():
    def FindNumsAppearOnce1(self, array):
        """
        Pythonic方式
        """
        if array == []:
            return []
        
        result = []
        for i in array:
            if array.count(i) == 1:
                result.append(i)
        
        return result

    def FindNumsAppearOnce2(self, array):
        """
        Pythonic方式，优化版
        """
        if array == []:
            return []
        
        return [i for i in array if array.count(i) == 1]

    def FindNumsAppearOnce3(self, array):
        """
        dict法
        """
        if array == []:
            return []
        
        dic = dict()
        for i in array:
            if not dic.get(i):
                dic[i] = 1
            else:
                dic[i] += 1
        
        return [i for i in array if dic[i] == 1]

    def FindNumsAppearOnce(self, array):
        """
        高效解法
        """
        if array == []:
            return []
        
        diff = 0
        for i in array:
            diff ^= i
        #diff &= (diff & 0xffffffff)
        diff &= -diff

        result = [0, 0]
        for i in array:
            if (i & diff) == 0: 
                result[0] ^= i
            else:
                result[1] ^= i
        return result

    def test(self, array):
        """
        测试函数
        """
        func_vec = [self.FindNumsAppearOnce1,
                    self.FindNumsAppearOnce2,
                    self.FindNumsAppearOnce3,
                    self.FindNumsAppearOnce]
        print("=====")
        for func in func_vec:
            start = time.time()
            result = func(array)
            end = time.time()
            print("result: {}, time(us): {:>5.2f}".format(result, (end - start)*10**6))


def main():
    array = [2, 4, 3, 6, 3, 2, 5, 5]
    array2 = [1, 2]

    s = Solution()
    s.test(array)
    s.test(array2)


if __name__ == "__main__":
    main()